Kuadrant/limitador-operator
# Limitador CRD new field: image

### What

Currently, the limitador image being used in the deployment is read from different sources with some order of precedence:
* If Limtador CR's `spec.version` is set -> image = `quay.io/kuadrant/limitador:${spec.version}` (note the repo is hardcoded)
* if `RELATED_IMAGE_LIMITADOR` env var is set -> image = `$RELATED_IMAGE_LIMITADOR`
* else: hardcoded to `quay.io/kuadrant/limitador:latest`

This PR introduces a new CRD field, `spec.image` that deprecates `spec.version` where you can set the full docker image URL. Allowing using custom docker image repos. Using private repos is not supported yet, we need to implement the pullSecret, another field where you can specify credentials to access private repos. 

The `spec.image` field is not meant to be used in production environments. It is meant to be used for dev/testing purposes. The main drawback of the `spec.image` usage is that upgrades cannot be supported as the limitador operator cannot ensure the operation to be safe. 

New order of precedence:
* If Limtador CR's `spec.image` is set -> image = `${spec.image}`
* If Limtador CR's `spec.version` is set -> image = `quay.io/kuadrant/limitador:${spec.version}` (note the repo is hardcoded)
* if `RELATED_IMAGE_LIMITADOR` env var is set -> image = `$RELATED_IMAGE_LIMITADOR`
* else: hardcoded to `quay.io/kuadrant/limitador:latest`


### Verification Steps
dev setup

```
make local-setup
```

Deploy the limitador CR with custom image URL (not from `quay.io/kuadrant/limitador`)

```yaml
k apply -f - <<EOF
---
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador
spec:
  image: quay.io/eastizle/limitador:custom-image-v1
EOF
```

Check deployment is correct

```
kubectl wait --timeout=300s --for=condition=Ready limitador limitador
```
Should return

```
limitador.limitador.kuadrant.io/limitador condition met
```

Check command line

```
kubectl get deployment/limitador-limitador  -o yaml | yq '.spec.template.spec.containers[0].image'
```
should look like this

```
quay.io/eastizle/limitador:custom-image-v1
```
